### DISCUSSION TITLE
Strategies for building cross-platform messaging apps?

### DISCUSSION THREAD

**Discussion by trosel** [markqvist/Reticulum#54] [Help & Questions]
It seems like all of the Reticulum SDKs are written in Python. The code is very clean and readable, but this presents a problem for those that would seek to build a cross-platform messaging app.

For example, I think the most popular cross-platform frameworks today are React Native and Flutter. How do we include [LXMF](https://github.com/markqvist/LXMF) into a React Native or Flutter application?

I have the following questions:

- Can we build a thin client that doesn't include any Python code and just uses Routing and Propagation Nodes to handle the messaging passing?
- Are there any Python libraries that allow you to build an app for Android and iOS both?
- What is the best way to compile a Python backend to be included on Android and iOS via something like React Native or Flutter?

**Comment by 4c3e** [comment]
Hello! This is a great question and one I am currently trying to work on right now. For context I have the far-off dream of potentially integrating reticulum as a transport layer for the messaging app [Briar](https://briarproject.org). Briar is a JVM based codebase so I am running into this same issue.

My current approach is something I call "[Reticulum Open Bridge](https://github.com/4c3e/ROB)" which is essentially a TCP sever that listens on a local socket, and relays data to and from RNS. This allows any programming language with a socket library to send and receive data over RNS, and makes writing new RNS libraries simple (with the downside that programs using these 'lite' libraries need to control the ROB process). This idea mirrors I2P's [SAM API](https://geti2p.net/en/docs/api/samv3).

Also worth noting there are frameworks for python that allow for apps that target Android, and Desktop (not sure about iOS support). There is a gui messenger using LXMF available for Android Linux and MacOS right now: https://github.com/markqvist/Sideband 

Eventually there is likely going to be some sort of RNS C implementation which could be packed into a binary and set up to be used in other apps like the current tor binaries. For example, Briar uses tor by executing and controlling a tor binary from within it's Java/Kotlin application.

**Reply by trosel** [reply]
There are many open source, cross-platform messaging apps built with React Native and Flutter at this point. It would be great if we could just swap the backend to use Reticulum instead.

> integrating reticulum as a transport layer for the messaging app [Briar](https://briarproject.org/).

The thing is, we should be able to use Reticulum (LXMF) as the messaging network *without* utilizing all of its features like LORA transport. 

If we ignore LORA and non TCP/IP based messaging, we can just use LXMF as a normal messenger that uses the regular internet. And then people that are in unique situations can still talk to those that are on the regular internet via LXMF. So if we ignore the fancy stuff, _it should make building a client easier_, no?

I don't think Briar is something we want to emulate because it doesn't support iOS, for example. There are multiple designers that would be willing to build clients on a more popular framework that has broader and deeper support in the tech community.

Sideband is a good example of the problem. It doesn't work on iOS (a massively huge OS that I and most of my friends happen to use).

**Reply by 4c3e** [reply]
>There are many open source, cross-platform messaging apps built with React Native and Flutter at this point. It would be great if we could just swap the backend to use Reticulum instead.

A similar ROB-type backend could be built specifically for LXMF. Then the flutter or react native app would just need to spawn this process in the background[1]. Once the process is running, the app can communicate over LXMF similar to normal socket programming.

Edit: I suppose a simple TCP client in dart or kotlin, that could connect to a LXMF Propagation node and sync messages with it would be another 'RNS lite' library idea. 

>Sideband is a good example of the problem. It doesn't work on iOS (a massively huge OS that I and most of my friends happen to use).

[1]: iOS unfortunately has major issues with allowing developers to use background process which makes building messaging apps that don't utilize apple's push notifications very hard ([more info here](https://code.briarproject.org/briar/briar/-/wikis/FAQ#will-there-be-an-ios-version-of-briar)). True p2p messaging apps can't perform well on iOS, but because LXMF has the propagation system, it might be possible to poll the network for messages every 15 minutes. But you wouldn't be able to receive messages in the background in real time (hopefully this changes soon).

**Reply by trosel** [reply]
> iOS unfortunately has major issues with allowing developers to use background process which makes building messaging apps that don't utilize apple's push notifications very hard ([more info here](https://code.briarproject.org/briar/briar/-/wikis/FAQ#will-there-be-an-ios-version-of-briar)).

This is how the Session app handles it. It gives users a choice. (apologies for the wall of text!) https://getsession.org/faq

> Session’s iOS client has two options for notifications: background polling (slow mode), and Apple Push Notification Service (APNs) (fast mode). 

> If you choose slow mode, the Session application runs in the background and periodically polls its swarm (see [What is a swarm](https://getsession.org/faq#swarms)) for new messages. If a new message is found, it is presented to you as a notification on your device.

> If you choose fast mode, Session will use APNs push notification service to deliver push notifications to your device. This requires your device IP address and unique push notification token are exposed to an Apple operated push notification server. Additionally, you will expose your Session ID and unique push notification token to an OPTF operated push notification server, for the purpose of providing notifications to the APNs server.

> These exposures are fairly minimal, because Apple will likely already know your device’s IP address through telemetry data or other applications on your device using push notifications. Registration of your Session ID and unique push notification token to the OPTF push notification server is necessary for detection and signaling of new messages and is low impact as registration occurs using onion requests meaning your Session ID and push notification token are never tied to any real world identifier (such as your IP address). 

> When using fast mode neither Apple nor the OPTF can see the contents of your messages, who you’re talking to, or exactly when messages are sent or received.

I just think that we shouldn't let "best" be the enemy of "good". There is really no "drag and drop" SDK today for p2p messaging (with forward privacy!). LXMF has the potential to be something amazing, but I think it is limited by the Python exclusivity. At least with my own knowledge, I don't know how to tell a front-end developer to use it in a client!

**Reply by 4c3e** [reply]
Also worth noting that [berty](berty.tech) is a messenger that functions similar to LXMF, it uses IPFS as a distributed message store, and polls for incoming messages. They have an experimental iOS app that apparently uses a custom notification system. Have not looked into specifics but it's something to keep an eye on.

**Reply by trosel** [reply]
I tried Berty 3 weeks ago and it was completely unusable. I think this has more to do with IPFS issues than anything else though.

**Reply by 4c3e** [reply]
@trosel same here, were you trying on iOS?

**Reply by trosel** [reply]
@4c3e yep I was using the iOS app. couldn't connect with my friend. connected with someone else and couldn't send messages. But wow the UI looks great :)

**Reply by 4c3e** [reply]
Yeah same experience for me on my iPhone, couldn't even get the bluetooth p2p working unfortunately. The project's still very early and they do have great front end work so far so it'll be exciting to watch them develop it further.

**Reply by trosel** [reply]
Back to this

> A similar ROB-type backend could be built specifically for LXMF. Then the flutter or react native app would just need to spawn this process in the background[1]. Once the process is running, the app can communicate over LXMF similar to normal socket programming.

> Edit: I suppose a simple TCP client in dart or kotlin, that could connect to a LXMF Propagation node and sync messages with it would be another 'RNS lite' library idea.

Do you have the knowledge to pursue this? what would it take? As we see with Berty, there is a distinct lack of workable solutions that Reticulum can fill. Biggest gap right now is a good multiplatform SDK (as I see it). If anyone can drag and drop LXMF in their user app and get messaging functionality for free, that is very attractive!

**Reply by markqvist** [reply]
LXMF will have a standalone daemon very soon, which means it will be much easier to interact with from other programs/ecosystems. I will *probably* have spooling of messages in the first release already, so you can just place outgoing messages in a filesystem directory, and have LXMF ingest and process them automatically. Same for incoming messages. Haven't worked out the specifics yet. Could also add some sort of JSON or msgpack api.

**Reply by 4c3e** [reply]
>Do you have the knowledge to pursue this? what would it take?

I definitely don't have the knowledge or the free time to work on mobile apps or some sort of fleshed-out SDK, unfortunately, I only have a decent-at-best understand of desktop app development. I think the best way forward for me is to continue spending weekends trying to build out proof-of-concept tools to help RNS developers that do have time for application development. 

>Haven't worked out the specifics yet. Could also add some sort of JSON or msgpack api.

Very cool. This coming weekend I'll try writing up a LXMF Open Bridge similar to ROB, and try and test it out with a small Kotlin application.
